### PR TITLE
refactor(blooms): Builder retrieves tasks from planner

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:

This PR adds the logic to the bloom builder to connect and receive tasks from the bloom planner. Tasks are not processed yet, only received.

**Special notes for your reviewer**:
- Retry mechanism s will be implemented on a different PR

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/bloombuild/builder/metrics.go [hunk 1]
```diff
@@ -22,5 +29,25 @@ func NewMetrics(r prometheus.Registerer) *Metrics {
...
+			Subsystem: metricsSubsystem,
+			Name:      "task_completed_total",
+			Help:      "Total number of task completed",
+		}, []string{"status"}),
+		taskTime: promauto.With(r).NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Subsystem: metricsSubsystem,
+			Name:      "task_time_seconds",
+			Help:      "Time spent processing a task.",
+			Buckets:   prometheus.DefBuckets,
```
Comment (chaudum) [inline pkg/bloombuild/builder/metrics.go:50]: The default bucket range might be too small.

File: pkg/bloombuild/builder/metrics.go [hunk 2]
```diff
@@ -22,5 +29,25 @@ func NewMetrics(r prometheus.Registerer) *Metrics {
...
+			Name:      "task_started_total",
+			Help:      "Total number of task started",
+		}),
+		taskCompleted: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Subsystem: metricsSubsystem,
+			Name:      "task_completed_total",
+			Help:      "Total number of task completed",
+		}, []string{"status"}),
+		taskTime: promauto.With(r).NewHistogramVec(prometheus.HistogramOpts{
```
Comment (chaudum) [inline pkg/bloombuild/builder/metrics.go:45]: nit: would name it taskDuration

**Review by chaudum — APPROVED** [review]

File: pkg/bloombuild/builder/builder.go [hunk 3]
```diff
@@ -2,49 +2,164 @@ package builder
...
+	if err := c.Send(&protos.BuilderToPlanner{BuilderID: b.ID}); err != nil {
+		return fmt.Errorf("failed to send ready message to planner: %w", err)
+	}
+
+	for b.State() == services.Running {
+		// When the planner connection closes or the builder stops, the context
+		// will be canceled and the loop will exit.
+		protoTask, err := c.Recv()
+		if err != nil {
+			if errors.Is(c.Context().Err(), context.Canceled) {
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/bloombuild/builder/builder.go:107]: If the connection context is canceled,  `c.Recv()` will return right after calling it.
(I actually had it as you suggest but noticed the behaviour while testing and decided to remove it for simplicity)

File: pkg/bloombuild/builder/metrics.go (hunk 1 above)
Comment (salvacorts) [reply pkg/bloombuild/builder/metrics.go]: Indeed. Changed

**Pull request merged**